app-backend/database/src/main/scala/com/azavea/rf/database/tables/Exports.scala

```diff
@@ -224,15 +232,45 @@ object Exports extends TableQuery(tag => new Exports(tag)) with LazyLogging {
       tool   <- OptionT(Tools.getTool(tRun.tool, user))
       ast    <- OptionT.pure[Future, MapAlgebraAST](tool.definition.as[MapAlgebraAST].valueOr(throw _))
       params <- OptionT.pure[Future, EvalParams](tRun.executionParameters.as[EvalParams].valueOr(throw _))
-      scenes <- OptionT(getScenes(export, user)
-        .map(_.map(s => s.ingestLocation.map(l => (s.id, l))).toList.sequence)
-      )
+      (scenes, projects) <- ingestLocs(params, user)
     } yield {
-      ASTInput(ast, params, scenes.toMap)
+      ASTInput(ast, params, scenes, projects)
     }
   }
 
+  /** Obtain the ingest locations for all Scenes and Projects which are
+    * referenced in the given [[EvalParams]]. If even a single Scene anywhere is
+    * found to have no `ingestLocation` value, the entire operation fails.
+    */
+  private def ingestLocs(
+    sources: EvalParams,
+    user: User
+  )(implicit database: DB): OptionT[Future, (Map[UUID, String], Map[UUID, List[(UUID, String)]])] = {
+
+    val (scenes, projects): (Stream[SceneRaster], Stream[ProjectRaster]) =
+      sources.sources.toStream
+        .map(_._2)
+        .foldLeft((Stream.empty[SceneRaster], Stream.empty[ProjectRaster]))({
+          case ((sacc, pacc), s: SceneRaster) => (s #:: sacc, pacc)
+          case ((sacc, pacc), p: ProjectRaster) => (sacc, p #:: pacc)
+        })
+
+    val scenesF: OptionT[Future, Map[UUID, String]] = scenes.map({ case SceneRaster(id, _) =>
+      OptionT(Scenes.getScene(id, user)).flatMap(s =>
+        OptionT.fromOption(s.ingestLocation.map((s.id, _)))
+      )
+    }).sequence.map(_.toMap)
```

Comment:
It's okay here. We're mapping into `OptionT` to get a `List[(UUID, List[(UUID, String)])]`, then `toMap` on the outer list. It's just a mapping from Project UUIDs to their *ordered* scenes, so the order of the outer list doesn't matter.